### PR TITLE
PostHog Metrics lib

### PR DESCRIPTION
**Pull request by gewfy**
- [x] `metrics.setConsent`
- [x] `metrics.logEvent`
- [x] `metrics.setUserProperties`
- [x] `metrics.setCoreProperties`
- [x] Enforce event and user properties types
- [x] Implement a couple of events and user properties
- [x] Updated `INSTALALTION.md`
- [x] Fix tests / mocks
- [x] Setup `production` and `staging` environments in PostHog
- [x] Repository secrets
- [ ] Invite everyone to PostHog

### CODE REVIEW

File: client/src/lib/metrics/types.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import {Session} from '../../../../shared/src/types/Session';
+
+export type Events = {
```
Comment (Nipher) [inline client/src/lib/metrics/types.ts:3]: types!!! Nice!

File: shared/src/types/UtilityTypes.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// credits goes to https://stackoverflow.com/a/50375286
+// function intersection producec - functin overloads
+export type UnionToIntersection<U> = (
```
Comment (Nipher) [inline shared/src/types/UtilityTypes.ts:3]: 🤯

File: client/src/lib/metrics/index.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+// Posthog event user properties https://posthog.com/docs/integrate/user-properties
+type EventUserProperties = {
+  $set?: AnyUserProperty;
+  $set_once?: AnyUserProperty;
+  $unset?: [keyof UserProperties];
+};
+
+// Generate all possible allowed event names and properties as logEvent overloads
+type LogEventOverloads = UnionToIntersection<
```
Comment (Nipher) [inline client/src/lib/metrics/index.tsx:49]: black magic right there hahaha
nice to get typed stuff on metrics

**Review by Nipher — APPROVED** [review]
Dope 🔥

File: shared/src/types/UtilityTypes.ts (hunk 2 above)
Comment (gewfy) [reply shared/src/types/UtilityTypes.ts:3]: Hahaha don't get me started on this 😂 But basically it was a deep ended hole getting a type object structure:
```javascript
 { eventName: { eventProp: type } }
```
Into be type checked in a function where they are separate arguments i.e. `logEvent(eventName, { eventProp: type });`.
What it does in the end is to create multiple function "overloads" of every possible combination:
```typescript
('some event name', { 'some event prop': string}) => void &
('some second event name', { 'some other event prop': boolean }) => void &
('some third event name', { 'some other event prop': boolean }) => void &
```
... and so forth...

Does it make sense? :D

File: client/src/lib/metrics/index.tsx (hunk 3 above)
Comment (gewfy) [reply client/src/lib/metrics/index.tsx:49]: Thanks! 🤗

File: client/src/lib/metrics/types.ts (hunk 1 above)
Comment (gewfy) [reply client/src/lib/metrics/types.ts:3]: Thanks! 🙏

**Pull request merged**